File: homework03-go/life_proto.go
package main

import (
	"github.com/hajimehoshi/ebiten"
	"image/color"
	"log"
	"math/rand"
	"time"
)

type GameOfLife struct {
	width      int
	height     int
	cellSize   int
	colsNumber int
	rowsNumber int
	grid       [][]bool
}

func (game *GameOfLife) init(width, height, cellSize int) {
	game.width = width
	game.height = height
	game.cellSize = cellSize
	game.colsNumber = game.width / game.cellSize
	game.rowsNumber = game.height / game.cellSize
	game.grid = game.createGrid(true)
}

func (game *GameOfLife) createGrid(randomize bool) [][]bool {
	grid := make([][]bool, game.rowsNumber, game.rowsNumber)
	for i := range grid {
		grid[i] = make([]bool, game.colsNumber, game.colsNumber)
	}

	if randomize {
		for i, row := range grid {
			for j := range row {
				grid[i][j] = rand.Intn(2) != 0
			}
		}
	}
	return grid
}

func (game *GameOfLife) drawLines(screen *ebiten.Image) {
	verticalLine, _ := ebiten.NewImage(1, game.height, ebiten.FilterDefault)
	_ = verticalLine.Fill(color.Black)

	horizontalLine, _ := ebiten.NewImage(game.width, 1, ebiten.FilterDefault)
	_ = horizontalLine.Fill(color.Black)

	for x := 0; x < game.width; x += game.cellSize {
		lineOpts := &ebiten.DrawImageOptions{}
		lineOpts.GeoM.Translate(float64(x), 0)
		_ = screen.DrawImage(verticalLine, lineOpts)
	}
	for y := 0; y < game.height; y += game.cellSize {
		lineOpts := &ebiten.DrawImageOptions{}
		lineOpts.GeoM.Translate(0, float64(y))
		_ = screen.DrawImage(horizontalLine, lineOpts)
	}
}

func (game *GameOfLife) drawGrid(screen *ebiten.Image) {
	cell, _ := ebiten.NewImage(game.cellSize-1, game.cellSize-1, ebiten.FilterDefault)
	_ = cell.Fill(color.RGBA{G: 255, A: 255})

	for y, row := range game.grid {
		for x, cellValue := range row {
			if cellValue {
				lineOpts := &ebiten.DrawImageOptions{}
				lineOpts.GeoM.Translate(float64(x*game.cellSize+1), float64(y*game.cellSize+1))
				_ = screen.DrawImage(cell, lineOpts)
			}
		}
	}
}

func (game *GameOfLife) run(screen *ebiten.Image) error {
	_ = screen.Fill(color.White)
	game.drawGrid(screen)
	game.drawLines(screen)
	game.grid = game.getNextGeneration()
	return nil
}

func (game *GameOfLife) getNeighboursCount(x, y int) int {
	count := 0
	for j := -1; j <= 1; j++ {
		for i := -1; i <= 1; i++ {
			if i == 0 && j == 0 { // Не рассматриваем текущую клетку
				continue
			}
			x1, y1 := x+i, y+j
			if x1 < 0 || y1 < 0 || game.colsNumber <= x1 || game.rowsNumber <= y1 { // Проверяем границы
				continue
			}
			if game.grid[y1][x1] {
				count++
			}
		}
	}
	return count
}

func (game *GameOfLife) getNextGeneration() [][]bool {
	nextGrid := game.createGrid(false)
	for y := 0; y < game.rowsNumber; y++ {
		for x := 0; x < game.colsNumber; x++ {
			neighbourNum := game.getNeighboursCount(x, y)
			if !game.grid[y][x] && neighbourNum == 3 {
				nextGrid[y][x] = true
			} else if game.grid[y][x] && (neighbourNum == 2 || neighbourNum == 3) {
				nextGrid[y][x] = true
			}
		}
	}
	return nextGrid
}

const (
	width  = 500
	height = 300
	size   = 10
)

func main() {
	rand.Seed(time.Now().UnixNano())

	game := &GameOfLife{}
	game.init(width, height, size)

	ebiten.SetRunnableInBackground(true)
	if err := ebiten.Run(game.run, width, height, 1, "Game of Life - Komarov Georgy, K3141"); err != nil {
		log.Fatal(err)
	}
}

File: homework03/life_proto.py
import random
from typing import List, Tuple

import pygame
from pygame.locals import *

Cell = Tuple[int, int]
Cells = List[int]
Grid = List[Cells]


class GameOfLife:
    def __init__(self, width: int = 640, height: int = 480, cell_size: int = 10, speed: int = 10) -> None:
        self.width = width
        self.height = height
        self.cell_size = cell_size

        # Устанавливаем размер окна
        self.screen_size = width, height
        # Создание нового окна
        self.screen = pygame.display.set_mode(self.screen_size)

        # Вычисляем количество ячеек по вертикали и горизонтали
        self.cell_width = self.width // self.cell_size
        self.cell_height = self.height // self.cell_size

        # Скорость протекания игры
        self.speed = speed

        # Создание списка клеток
        self.grid = self.create_grid()

    def draw_lines(self) -> None:
        """ Отрисовать сетку """
        for x in range(0, self.width, self.cell_size):
            pygame.draw.line(self.screen, pygame.Color('black'), (x, 0), (x, self.height))
        for y in range(0, self.height, self.cell_size):
            pygame.draw.line(self.screen, pygame.Color('black'), (0, y), (self.width, y))

    def draw_grid(self) -> None:
        """
        Отрисовка списка клеток с закрашиванием их в соответствующе цвета.
        """
        for y in range(len(self.grid)):
            for x in range(len(self.grid[0])):
                if self.grid[y][x]:
                    pygame.draw.rect(self.screen,
                                     pygame.Color('green'),
                                     (x * self.cell_size + 1,
                                      y * self.cell_size + 1,
                                      self.cell_size - 1,
                                      self.cell_size - 1)
                                     )

    def run(self) -> None:
        """ Запустить игру """
        pygame.init()
        clock = pygame.time.Clock()
        pygame.display.set_caption('Game of Life')
        self.screen.fill(pygame.Color('white'))

        running = True
        while running:
            for event in pygame.event.get():
                if event.type == QUIT:
                    running = False
            self.screen.fill(pygame.Color('white'))
            self.draw_lines()
            self.draw_grid()

            # Отрисовка списка клеток
            # Выполнение одного шага игры (обновление состояния ячеек)
            self.grid = self.get_next_generation()
            pygame.display.flip()
            clock.tick(self.speed)
        pygame.quit()

    def create_grid(self, randomize: bool = False) -> Grid:
        """
        Создание списка клеток.

        Клетка считается живой, если ее значение равно 1, в противном случае клетка
        считается мертвой, то есть, ее значение равно 0.

        Parameters
        ----------
        randomize : bool
            Если значение истина, то создается матрица, где каждая клетка может
            быть равновероятно живой или мертвой, иначе все клетки создаются мертвыми.

        Returns
        ----------
        out : Grid
            Матрица клеток размером `cell_height` х `cell_width`.
        """
        grid = [[0 for _ in range(self.cell_width)] for _ in range(self.cell_height)]
        if randomize:
            for y in range(self.cell_height):
                for x in range(self.cell_width):
                    grid[y][x] = random.randint(0, 1)
        return grid

    def get_neighbours(self, cell: Cell) -> Cells:
        """
        Вернуть список соседних клеток для клетки `cell`.

        Соседними считаются клетки по горизонтали, вертикали и диагоналям,
        то есть, во всех направлениях.

        Parameters
        ----------
        cell : Cell
            Клетка, для которой необходимо получить список соседей. Клетка
            представлена кортежем, содержащим ее координаты на игровом поле.

        Returns
        ----------
        out : Cells
            Список соседних клеток.
        """
        y0, x0 = cell

        coords = []
        for x in range(-1, 2):
            for y in range(-1, 2):
                x1, y1 = x0 + x, y0 + y
                if (x or y) and 0 <= x1 < self.cell_width and 0 <= y1 < self.cell_height:
                    coords.append((y1, x1))

        return [self.grid[y][x] for y, x in coords]

    def get_next_generation(self) -> Grid:
        """
        Получить следующее поколение клеток.

        Returns
        ----------
        out : Grid
            Новое поколение клеток.
        """
        next_grid = self.create_grid()
        for y in range(self.cell_height):
            for x in range(self.cell_width):
                neighbours_n = self.get_neighbours((y, x)).count(1)
                if not self.grid[y][x] and neighbours_n == 3:
                    next_grid[y][x] = 1
                elif self.grid[y][x] and neighbours_n in [2, 3]:
                    next_grid[y][x] = 1
        return next_grid


if __name__ == '__main__':
    game = GameOfLife()
    game.grid = game.create_grid(randomize=True)

    game.run()
